chain/client/src/sync/state/shard.rs

```diff
@@ -0,0 +1,268 @@
+                part_id,
+                header.clone(),
+                cancel.clone(),
+            );
+            respawn_for_parallelism(&*future_spawner, "state sync download part", future)
+        })
+        .buffer_unordered(MAX_PARALLELISM_PER_SHARD_FOR_FAIRNESS)
+        .try_collect::<Vec<_>>()
+        .await?;
+
+    return_if_cancelled!(cancel);
+    *status.lock().unwrap() = ShardSyncStatus::StateApplyInProgress;
+    runtime.get_tries().unload_mem_trie(&shard_uid);
+    let mut store_update = store.store_update();
+    runtime
+        .get_flat_storage_manager()
+        .remove_flat_storage_for_shard(shard_uid, &mut store_update.flat_store_update())?;
+    store_update.commit()?;
+
+    return_if_cancelled!(cancel);
+    tokio_stream::iter(0..num_parts)
+        .map(|part_id| {
+            let store = store.clone();
+            let runtime = runtime.clone();
+            let computation_task_tracker = computation_task_tracker.clone();
+            let cancel = cancel.clone();
+            let future = apply_state_part(
+                store,
+                runtime,
+                computation_task_tracker,
+                cancel,
+                sync_hash,
+                shard_id,
+                part_id,
+                num_parts,
+                state_root,
+                epoch_id,
+            );
+            respawn_for_parallelism(&*future_spawner, "state sync apply part", future)
+        })
+        .buffer_unordered(MAX_PARALLELISM_PER_SHARD_FOR_FAIRNESS)
+        .try_collect::<Vec<_>>()
+        .await?;
+
+    return_if_cancelled!(cancel);
+    // Create flat storage.
+    {
+        let (epoch_id, _) = query_epoch_id_and_height_for_block(&store, sync_hash)?;
+        let shard_uid = epoch_manager.shard_id_to_uid(shard_id, &epoch_id)?;
+        let chunk = header.cloned_chunk();
+        let block_hash = chunk.prev_block();
+
+        // We synced shard state on top of _previous_ block for chunk in shard state header and applied state parts to
+        // flat storage. Now we can set flat head to hash of this block and create flat storage.
+        // If block_hash is equal to default - this means that we're all the way back at genesis.
+        // So we don't have to add the storage state for shard in such case.
+        // TODO(8438) - add additional test scenarios for this case.
+        if *block_hash != CryptoHash::default() {
+            create_flat_storage_for_shard(&store, &*runtime, shard_uid, &chunk)?;
+        }
+    }
+    return_if_cancelled!(cancel);
+    // Load memtrie.
+    {
+        let handle = computation_task_tracker.get_handle(&format!("shard {}", shard_id)).await;
+        handle.set_status("Loading memtrie");
+        runtime.get_tries().load_mem_trie_on_catchup(&shard_uid, &state_root)?;
+    }
+
+    return_if_cancelled!(cancel);
+
+    // Finalize; this needs to be done by the Chain.
+    *status.lock().unwrap() = ShardSyncStatus::StateApplyFinalizing;
```

Comment:
Nice! With this setup, it should ideally be possible for us to convert ShardSyncStatus to a string instead? And we can keep adding more and more status with more information?

Reply:
I didn't want to touch that part because it was also used in a couple of other places. Maybe I can clean it up after.